File: AppSets-iOS/AppSets/com/app/appsets/im/ImMessageGenerator.swift
//
//  ImMessageGenerator.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/9.
//

import Foundation
import RMQClient.RMQMessage


class ImMessageGenerator {
    
    private static let TAG = "ImMessageGenerator"
    
    let decoder = JSONDecoder()
    let encoder = JSONEncoder()
    
    public static let Instance = ImMessageGenerator()
    
    private init() {
        
    }
    
    func generateBySend(
        context: Context,
        session: Session,
        inputSelector: InputSelector,
        content: Any
    ) -> any ImMessage {
        PurpleLogger.current.d(ImMessageGenerator.TAG, "generateBySend")
        let messageId = UUID().uuidString
        let timestamp = Date()
        let fromUserInfo = LocalAccountManager.Instance.userInfo
        let messageFromInfo = MessageFromInfo(
            uid: fromUserInfo.uid,
            name: fromUserInfo.name,
            avatarUrl: fromUserInfo.avatarUrl
        )
        
        let messageToInfo = MessageToInfo.fromImObj(session.imObj)
        
    
        var messageContent = ""
        if content is String {
            messageContent = content as! String
        }
        var metadata: StringMessageMetadata = StringMessageMetadata(
            description: "", size: 0, compressed: false, encode: "none", data: messageContent, contentType: ContentType.APPLICATION_TEXT
        )
        
        let imMessage = IM_Text(
            id: messageId,
            timestamp: timestamp,
            fromInfo: messageFromInfo,
            toInfo: messageToInfo,
            messageGroupTag: nil,
            metadata: metadata
        )
        return imMessage
    }
    
    func generateByReceived(
        _ message: RMQMessage
    ) -> (any ImMessage)? {
        let readableContent = String(data: message.body, encoding: .utf8) ?? ""
        PurpleLogger.current.d(
            ImMessageGenerator.TAG,
            "generateByReceived, message:\(readableContent)"
        )
        
        let headers = message.headers()
        
        let imMessageType = message.messageType()
        
        let imMessageTimestamp = message.timestamp()
        
        guard let fromUid = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_UID
        }) else {
            return nil
        }
        
        guard let messageId = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_ID
        }) else {
            return nil
        }
        
        guard let toType = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_TYPE
        }) else {
            return nil
        }
        let fromName = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_NAME
        })
        
        let fromNameBase64 = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_NAME_BASE64
        })
        
        let fromAvatarUrl = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_AVATAR_URL
        })
        
        let fromRoles = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_ROLES
        })

        let groupMessageTag = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_MESSAGE_GROUP_TAG
        })
        
        let toId = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_ID
        })
        
        let toName = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_NAME
        })
        
        let toNameBase64 = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_NAME_BASE64
        })
        
        let toIconUrl = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_ICON_URL
        })
        
        let toRoles = headers?.first(where: { (key: String, value: NSObject) in
            key == ImMessageConstant.HEADER_MESSAGE_TO_ROLES
        })
        
        let messageFromInfo = MessageFromInfo(
            uid: getDicStringValue(fromUid),
            name: getDicStringValueUnWrappeBase64(fromNameBase64),
            avatarUrl: getDicStringValue(fromAvatarUrl),
            roles: getDicStringValue(fromRoles)
        )
        
        let messageToInfo = MessageToInfo(
            toType: getDicStringValue(toType),
            id: getDicStringValue(toId),
            name: getDicStringValueUnWrappeBase64(toNameBase64),
            iconUrl: getDicStringValue(toIconUrl),
            roles: getDicStringValue(toRoles)
        )
        
        let timestamp = imMessageTimestamp ?? Date()
        
        let metadataJsonString = String(data: message.body, encoding: .utf8) ?? ""
        
        let metadataJsonStringData = metadataJsonString.data(using: .utf8)!
        let metadata = try! decoder.decode(StringMessageMetadata.self, from: metadataJsonStringData)
        
        let imMessage = IM_Text(
            id: getDicStringValue(messageId),
            timestamp: timestamp,
            fromInfo: messageFromInfo,
            toInfo: messageToInfo,
            messageGroupTag: getDicStringValue(groupMessageTag),
            metadata: metadata
        )
        
        return imMessage
    }
    
    func generateByLocalDB(
    ) -> (any ImMessage)? {
        PurpleLogger.current.d(ImMessageGenerator.TAG, "generateByLocalDB")
        return nil
    }
    
    private func getDicStringValueUnWrappeBase64(_ element: Dictionary<String, NSObject>.Element?) -> String {
        let base64String = getDicStringValue(element)
        let decodedString = String.unWrappeBase64(base64String) ?? base64String
        return decodedString
    }
    
    private func getDicStringValue(_ element: Dictionary<String, NSObject>.Element?) -> String {
        PurpleLogger.current.d(ImMessageGenerator.TAG, "getDicStringValue, key:\(String(describing: element?.key))")
        if element == nil {
            return ""
        }
        return getNSObjStringValue(element!.value)
    }
    
    private func getNSObjStringValue(_ obj: NSObject) -> String {
        PurpleLogger.current.d(ImMessageGenerator.TAG, "getNSObjStringValue, obj:\(obj)")
        if obj is RMQLongstr {
            let valueLongString = (obj as! RMQLongstr)
            return valueLongString.stringValue
        }
        if obj is RMQVoid {
            return ""
        }
        if obj is NSString {
            return (obj as! NSString).description
        }
        if obj is Foundation.Data {
            return String(data:(obj as! Foundation.Data), encoding: .utf8) ?? ""
        }
        
        do{
            let value = try (obj.value(forKey: "stringValue") as? String) ?? ""
            return value
        }catch let error {
            return ""
        }
    }
    
    func makeMessageMetadataAsJsonStringData(message: any ImMessage)-> Data? {
        PurpleLogger.current.d(ImMessageGenerator.TAG, "makeMessageMetadataAsJsonString, for messageId:\(message.id)")
        return try? encoder.encode(message.metadata)
    }
    
}

File: AppSets-Servers-Spring/MainServer/src/main/kotlin/xcj/app/main/im/ImMessageGenerator.kt
package xcj.app.main.im

import com.google.gson.Gson
import com.rabbitmq.client.AMQP
import com.rabbitmq.client.LongString
import xcj.app.main.model.res.UserInfoRes
import xcj.app.main.model.table.mysql.appSetsUserAdmin0
import xcj.app.main.util.ContentType
import xcj.app.util.PurpleLogger
import java.util.*

object ImMessageGenerator {

    private const val TAG = "ImMessageGenerator"

    val gson = Gson()

    private val messageFromInfo by lazy {
        val appSetsUserAdmin0 = appSetsUserAdmin0()
        MessageFromInfo(
            appSetsUserAdmin0.uid!!,
            appSetsUserAdmin0.name!!,
            appSetsUserAdmin0.avatarUrl!!,
            MessageToInfo.ROLE_ADMIN
        )
    }

    fun generateBySend(
        toUserInfo: UserInfoRes,
        content: Any
    ): ImMessage {
        PurpleLogger.current.d(TAG, "generateBySend")
        val messageToInfo =
            MessageToInfo(
                MessageToInfo.TO_TYPE_O2O,
                toUserInfo.uid,
                toUserInfo.name,
                toUserInfo.avatarUrl,
                null
            )
        val messageContent = gson.toJson(content)
        val metadata = ImMessage.StringMessageMetadata(
            "system_message",
            0,
            false,
            "none",
            messageContent,
            ContentType.TEXT_PLAIN
        )

        return ImMessage.System(
            UUID.randomUUID().toString(),
            Date(),
            messageFromInfo,
            messageToInfo,
            null,
            metadata,
        )
    }

    fun generateByReceived(properties: AMQP.BasicProperties?, body: ByteArray?): ImMessage? {
        val headers = properties?.headers ?: run {
            PurpleLogger.current.d(TAG, "generateByReceived, headers is null, return")
            return null
        }

        val fromUid = if (headers.containsKey(ImMessage.HEADER_MESSAGE_UID)) {
            (headers[ImMessage.HEADER_MESSAGE_UID] as? LongString).toString()
        } else {
            return null
        }
        val messageId = if (headers.containsKey(ImMessage.HEADER_MESSAGE_ID)) {
            (headers[ImMessage.HEADER_MESSAGE_ID] as? LongString).toString()
        } else {
            return null
        }
        val fromName = if (headers.containsKey(ImMessage.HEADER_MESSAGE_NAME)) {
            (headers[ImMessage.HEADER_MESSAGE_NAME] as? LongString).toString()
        } else {
            null
        }
        val fromAvatarUrl = if (headers.containsKey(ImMessage.HEADER_MESSAGE_AVATAR_URL)) {
            (headers[ImMessage.HEADER_MESSAGE_AVATAR_URL] as? LongString).toString()
        } else {
            null
        }
        val fromRoles = if (headers.containsKey(ImMessage.HEADER_MESSAGE_ROLES)) {
            (headers[ImMessage.HEADER_MESSAGE_ROLES] as? LongString).toString()
        } else {
            null
        }
        val messageGroupTag = if (headers.containsKey(ImMessage.HEADER_MESSAGE_MESSAGE_GROUP_TAG)) {
            (headers[ImMessage.HEADER_MESSAGE_MESSAGE_GROUP_TAG] as? LongString).toString()
        } else {
            null
        }
        val toType = if (headers.containsKey(ImMessage.HEADER_MESSAGE_TO_TYPE)) {
            (headers[ImMessage.HEADER_MESSAGE_TO_TYPE] as? LongString).toString()
        } else {
            return null
        }
        val toId = if (headers.containsKey(ImMessage.HEADER_MESSAGE_TO_ID)) {
            (headers[ImMessage.HEADER_MESSAGE_TO_ID] as? LongString).toString()
        } else {
            return null
        }
        val toName = if (headers.containsKey(ImMessage.HEADER_MESSAGE_TO_NAME)) {
            (headers[ImMessage.HEADER_MESSAGE_TO_NAME] as? LongString).toString()
        } else {
            return null
        }
        val toRoles = if (headers.containsKey(ImMessage.HEADER_MESSAGE_TO_ROLES)) {
            (headers[ImMessage.HEADER_MESSAGE_TO_ROLES] as? LongString).toString()
        } else {
            null
        }
        val toIconUrl = if (headers.containsKey(ImMessage.HEADER_MESSAGE_TO_ICON_URL)) {
            (headers[ImMessage.HEADER_MESSAGE_TO_ICON_URL] as? LongString).toString()
        } else {
            null
        }

        val messageFromInfo = MessageFromInfo(fromUid, fromName, fromAvatarUrl, fromRoles)

        val messageToInfo = MessageToInfo(toType, toId, toName, toIconUrl, toRoles)

        val messageType = properties.type

        val messageMetadataString = body?.decodeToString() ?: ""

        val timestamp = properties.timestamp
        val imMessage = generateImMessage(
            messageType,
            messageMetadataString,
            messageId,
            timestamp,
            messageFromInfo,
            messageToInfo,
            messageGroupTag
        )
        return imMessage
    }

    private fun generateImMessage(
        messageType: String,
        messageMetadataString: String,
        messageId: String,
        timestamp: Date,
        messageFromInfo: MessageFromInfo,
        messageToInfo: MessageToInfo,
        messageGroupTag: String?
    ): ImMessage? {
        return when (messageType) {
            ImMessageDesignType.TYPE_TEXT -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.Text(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_IMAGE -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.Image(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }


            ImMessageDesignType.TYPE_VOICE -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.Voice(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_LOCATION -> {
                val messageMetadata = gson.fromJson<ImMessage.LocationMessageMetadata>(
                    messageMetadataString,
                    ImMessage.LocationMessageMetadata::class.java
                )
                ImMessage.Location(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_VIDEO -> {
                val messageMetadata = gson.fromJson<ImMessage.VideoMessageMetadata>(
                    messageMetadataString,
                    ImMessage.VideoMessageMetadata::class.java
                )
                ImMessage.Video(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_AD -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.Ad(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }


            ImMessageDesignType.TYPE_MUSIC -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.Music(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_FILE -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.File(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            ImMessageDesignType.TYPE_HTML -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.HTML(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }


            ImMessageDesignType.TYPE_SYSTEM -> {
                val messageMetadata = gson.fromJson<ImMessage.StringMessageMetadata>(
                    messageMetadataString,
                    ImMessage.StringMessageMetadata::class.java
                )
                ImMessage.System(
                    messageId, timestamp,
                    messageFromInfo, messageToInfo,
                    messageGroupTag, messageMetadata
                )
            }

            else -> null
        }
    }

    fun makeMessageMetadataAsJsonString(message: ImMessage): String {
        PurpleLogger.current.d(TAG, "makeMessageMetadataAsJsonString, for messageId:${message.id}")
        return gson.toJson(message.metadata)
    }
}